apps/extension/src/storage/base.ts

```diff
@@ -14,75 +15,71 @@ export interface IStorage {
   };
 }
 
-export interface StorageItem<T> {
-  version: string;
-  value: T;
-}
+export type MigrationFn<OldState, NewState> = (prev: OldState) => NewState | Promise<NewState>;
 
-type Version = string;
+export type ExtensionStorageDefaults<T extends { dbVersion: number }> = Required<
+  Omit<T, 'dbVersion'>
+>;
 
-export type MigrationMap<OldState, NewState> = {
-  [K in keyof OldState & keyof NewState]?: (
-    prev: OldState[K],
-    get: <K extends keyof NewState>(key: K) => Promise<NewState[K]>,
-  ) => NewState[K] | Promise<NewState[K]>;
-};
+export interface RequiredMigrations {
+  // Explicitly require key '0' representing...
+  // It is quite difficult writing a generic that covers all migration function kinds.
+  // Therefore, the writer of the migration should ensure it is typesafe when they define it.
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  0: MigrationFn<any, any>;
```

Comment:
should we make this an _optional_ field? for instance, migration function for version 0 won't be relevant for a higher version.